### DISCUSSION TITLE
Performance bottlenecks and findings

### DISCUSSION THREAD

**Discussion by Boshen** [web-infra-dev/rspack#1586] [General]
A summary of perf bottlenecks and findings.

General background research study:

follow the work of https://github.com/nnethercote, read his [blog](https://nnethercote.github.io/), read the [perf book](https://nnethercote.github.io/perf-book/title-page.html)

**Comment by Boshen** [comment]
| Summary | Issue / PR    | Prod | Dev |
| ------- | ----- | ---- | --- |
| Use precomputed hash in `HashMap<Ustr, _>` | #1567 #1568 | 6% | 3%
| Remove extra source code clone in javascript plugin | #1571 #1579 | 12%  | 2%
| Enable concurrent mode in swc minifier | #1583 | 13%     | -
| Use rustc_hash::FxHasher | #1589 #1595 | 5%     | 5%
| cache module identifier for NormalModule | #1636 | 5%     | 2%
| use String::from_utf8_unchecked for codegen | #1648 | 5%     | 5%
| use BufWriter for file write  | #1657 | 2%     | 2%

**Comment by Boshen** [comment]
# [nodejs_resolver](https://github.com/modern-js-dev/nodejs_resolver)

See the [list of PRs](https://github.com/modern-js-dev/nodejs_resolver/pulls?q=is%3Apr+is%3Aclosed+author%3ABoshen)

Overall we (with @bvanjoi) have reduced the application time by a constant of 1 to 2 second, and ~10% for large applications.

* Reduce syscall
  * Removed `std::fs::canonicalize`, which is an expensive syscall.
  * Reduced `std::fs::metadata` stat call by half
  * Reduced cpu time of doing syscalls by ~15 times
  * Removed useless stat calls on package.json, e.g. `test.js/package.json`
* Reduce memory usage
  * Removed all unnecessary `PathBuf` allocations
* Reduce CPU usage
  * Removed unnecessary usage of `SmolStr`
  * Added rustc_hash::FxHasher for faster hashing

<div>
<img src="https://user-images.githubusercontent.com/1430279/218914353-a2dacc51-f18f-4e04-ae32-e3c88cf7311d.png" alt="drawing" width="45%" />
<img src="https://user-images.githubusercontent.com/1430279/218914947-36b12b45-8297-4213-b02f-c73c1a288cac.png" alt="drawing" width="45%" />
<div>

**Comment by Boshen** [comment]
When writing file to disk, the string is currently joined together all at once, which creates a huge memory allocation pressure.

https://github.com/modern-js-dev/rspack/blob/45c912cc0c5a3b5aa2d9fa5c721638c6ae647a4b/crates/rspack_core/src/compiler/mod.rs#L205

https://github.com/modern-js-dev/rspack-sources/blob/1969fb3d48fdab5bbb4d0a3b81a8aa87e1867954/src/concat_source.rs#L85-L93 

Instead, we should just iterate the sources and write it into a BufferWriter to balance the performance around memory allocation and the `fs::write` syscall.

**Comment by Boshen** [comment]
# Memory Footprint

## 4.5kiB

### Clone source before parse

<img width="1726" alt="image" src="https://user-images.githubusercontent.com/1430279/223033996-fe74ce0c-ed45-446e-bb6a-8fa265e62bcc.png">

https://github.com/web-infra-dev/rspack/blob/20f0fd9ea5daa9f99c4743dae50e1c565d754b4b/crates/rspack_plugin_javascript/src/plugin.rs#L260-L265

Why do we need to clone the whole source before parsing? It should be able to take a reference instead.


## 2.5KiB
This is really heavy:

<img width="2483" alt="image" src="https://user-images.githubusercontent.com/1430279/223348066-997245ed-cd7c-4b8f-98bb-f3a5a685fdfd.png">

## 112 Bytes

There's a very frequent allocation here somewhere, but I can't find it.

<img width="2282" alt="image" src="https://user-images.githubusercontent.com/1430279/223313395-5a51e30e-75b0-45fa-8992-94c285bb5b76.png">

## 18.86 MiB

<img width="1637" alt="image" src="https://user-images.githubusercontent.com/1430279/223348857-65e2d007-49a5-4746-8d56-9236c6b8f62e.png">


----

It feels like AST is not dropped at all.

![8eb5e08b-ca8d-46a8-b3bf-6dfced66a5f1](https://user-images.githubusercontent.com/1430279/223346292-b7ba35bf-ebec-4d04-9a7d-6618179f1f31.jpeg)

**Reply by hyf0** [reply]
I look into the SWC.

It looks like SWC requires mutable `String` to remove possible `bom`.

https://github.com/swc-project/swc/blob/66247f0c49e890aca94d1788be41e1d36d21ed2f/crates/swc_common/src/syntax_pos.rs#L879

**Reply by kdy1** [reply]
I made https://github.com/swc-project/swc/pull/7029/files
You can remove bom earlier and store it as `Arc<String>`

**Comment by Boshen** [comment]
# SWC 

SWC AST is not fine tuned, for example:

## Expression

Is 112 bytes, which puts a lot of allocation pressure all over the place.

<img width="593" alt="image" src="https://user-images.githubusercontent.com/1430279/223063638-32990bc8-4df2-4677-9583-42f25fd909f0.png">

Some of the enums can be shrink by boxing their attributes, it would be nice if we can get `Expr` below 64 bytes. This allows applies for `Stmt`.

## Reallocate in `parse_block_body`

<img width="1533" alt="image" src="https://user-images.githubusercontent.com/1430279/223035289-4aff8790-00c3-42a2-ba98-74af75fc89e0.png">

https://github.com/swc-project/swc/blob/f32be0e7c60ade1288e27731996de428b957dff8/crates/swc_ecma_parser/src/parser/stmt.rs#L72

https://github.com/swc-project/swc/blob/f32be0e7c60ade1288e27731996de428b957dff8/crates/swc_ecma_parser/src/parser/stmt.rs#L29

https://github.com/thomcc/rust-typed-arena/blob/a6abfd34491ba7bdb72255c853ba6fdccffac112/src/lib.rs#L433-L445

## Allocate all statements

<img width="1338" alt="image" src="https://user-images.githubusercontent.com/1430279/223036295-aca0325e-5ade-4e56-bd95-c442812cae47.png">

https://github.com/swc-project/swc/blob/f32be0e7c60ade1288e27731996de428b957dff8/crates/swc_ecma_transforms_optimization/src/simplify/branch/mod.rs#L1293


## Short lived 16 bytes in `read_number`

<img width="2301" alt="image" src="https://user-images.githubusercontent.com/1430279/223351820-6c5cbb58-6182-42b1-81ee-f02062bebdf2.png">

**Reply by Boshen** [reply]
I wrote an issue to swc https://github.com/swc-project/swc/issues/7019

**Reply by kdy1** [reply]
The `parse_block_body` issue will be fixed if we box more statements

**Reply by kdy1** [reply]
Actually I considered it in the past, but didn't do it because I have other tasks to work on

**Comment by Boshen** [comment]
# CPU

<img width="2149" alt="image" src="https://user-images.githubusercontent.com/1430279/223396352-1bb25e62-09ce-4720-b5ab-8b265717db6a.png">

This is blocking everything for 500ms.

**Reply by hyf0** [reply]
It should be improved after merging https://github.com/swc-project/swc/pull/7030.

**Comment by Boshen** [comment]
# CPU Utilization

<img width="1921" alt="image" src="https://user-images.githubusercontent.com/1430279/223679553-16362ecd-d684-4815-ac10-d6ea20900ac5.png">
